### PR TITLE
[RFC] shada.c: Fix HAVE_BE64TOH check

### PR DESCRIPTION
**Pull request by ghost**
mentioned here: https://github.com/neovim/neovim/pull/3985#issuecomment-170663426

---

HAVE_BE64TOH is defined in config/config.h, which is included by vim.h.  Since the HAVE_BE64TOH check in shada.c is evaluated before vim.h is included, it always evaluates to false, meaning that be64toh() in shada.c is always used instead of the one in `<endian.h>`.  

Moving the HAVE_BE64TOH block to after where vim.h is included seems to fix the issue.

### CODE REVIEW

File: src/nvim/shada.c [hunk 1]
```diff
@@ -10,9 +5,6 @@
 #include <stdint.h>
 #include <inttypes.h>
 #include <errno.h>
-#ifdef HAVE_UNISTD_H
-# include <unistd.h>
-#endif
```
Comment (Pyrohh) [inline src/nvim/shada.c]: This is already included for non-Windows systems in `os/os_defs.h`, which is included by `ascii.h`, which is included by `shada.c`.

**Comment by justinmk** [comment]
LGTM.

File: src/nvim/shada.c (hunk 1 above)
Comment (ZyX-I) [inline src/nvim/shada.c]: `ascii.h` including `os/*` stuff is not reliable. If correct unistd include is os_defs.h, it should be here explicitly.

File: src/nvim/shada.c [hunk 2]
```diff
@@ -21,6 +13,11 @@
 #include "nvim/os/os.h"
 #include "nvim/os/time.h"
 #include "nvim/vim.h"
+#ifdef HAVE_BE64TOH
+# define _BSD_SOURCE 1
+# define _DEFAULT_SOURCE 1
+# include <endian.h>
+#endif
```
Comment (ZyX-I) [inline src/nvim/shada.c]: And I would say it is better placed after `nvim` defines list, not break this large group into pieces. Or be in the same group as `vim.h` above: `vim.h+endian.h`, blank, `os/os.h+…+lib/kvec.h`.

File: src/nvim/shada.c (hunk 1 above)
Comment (ZyX-I) [inline src/nvim/shada.c]: (And you did not mention this change in the commit message: it talks only about HAVE_BE64TOH.)

File: src/nvim/shada.c (hunk 2 above)
Comment (Pyrohh) [inline src/nvim/shada.c]: Good idea, I put it after the local includes.

File: src/nvim/shada.c (hunk 1 above)
Comment (Pyrohh) [inline src/nvim/shada.c]: :+1: I just reverted that change. It seems excessive to include a shared header such as `os/os_defs.h` just for one system header.

**Pull request merged**